File: workspace-flink/src/weaklyConnectedComponents/src/main/java/jar/weaklyConnectedComponents.java
package jar;

import java.io.File;
import java.io.FileWriter;
import java.util.ArrayList;
import java.io.BufferedWriter;
import org.apache.flink.graph.Edge;
import org.apache.flink.graph.Graph;
import org.apache.flink.graph.Vertex;
import org.apache.flink.types.NullValue;
import org.apache.flink.api.java.tuple.*;
import org.apache.flink.api.java.DataSet;
import org.apache.flink.api.common.typeinfo.Types;
import org.apache.flink.api.java.ExecutionEnvironment;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.graph.library.ConnectedComponents;

public class weaklyConnectedComponents {

	public static void main(String[] args) throws Exception {
		
		// Setup the execution environment
		final ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();

		String edgeListFilePath = "/home/user/data/web-Google.txt";

		// Load the edges as a DataSet of Tuples
		DataSet<Tuple2<Integer, Integer>> edges = env.readTextFile(edgeListFilePath)
			.filter(line -> !line.startsWith("#"))
			.map(line -> {
				String[] tokens = line.split("\\s+");
				return new Tuple2<Integer, Integer>(Integer.parseInt(tokens[0]), Integer.parseInt(tokens[1]));
			}).returns(Types.TUPLE(Types.INT, Types.INT));

		// Create the graph
		Graph<Integer, NullValue, NullValue> graph = Graph.fromTuple2DataSet(edges, env);

		// Maximum number of iterations, big enough to converge
		int iters = 100;

		// Annotate each vertex with its ID as a value
		Graph<Integer, Integer, NullValue> annotatedGraph = graph.mapVertices(new MapFunction<Vertex<Integer, NullValue>, Integer>() {
			public Integer map(Vertex<Integer, NullValue> value) {
				return value.getId();
			}
		});
		
		// Compute the weakly connected components
		ConnectedComponents<Integer, Integer, NullValue> connectedComponents = new ConnectedComponents<>(iters);
		DataSet<Vertex<Integer, Integer>> result = connectedComponents.run(annotatedGraph);

		ArrayList<Vertex<Integer, Integer>> components = new ArrayList<Vertex<Integer, Integer>>();
		result.collect().forEach(components::add);

		// Write the output to a file
		File outputs = new File("/home/user/workspace-flink/outputs/weaklyConnectedComponents.txt");
		BufferedWriter bw = new BufferedWriter(new FileWriter(outputs));
		for (Vertex<Integer, Integer> vertex : components) {
			bw.write(vertex.getId() + " " + vertex.getValue() + "\n");
		}
		bw.close();
	}
}

File: workspace-graphx/src/main/scala/weaklyConnectedComponents.scala
import java.io._
import org.apache.spark._
import org.apache.spark.rdd.RDD
import org.apache.spark.graphx._
import org.apache.spark.SparkConf
import org.apache.spark.SparkContext
import org.apache.spark.storage.StorageLevel

object weaklyConnectedComponents {
    def main(args: Array[String]): Unit = {
        
        // Create a Spark configuration and context
        val conf = new SparkConf()
            .setAppName("weaklyConnectedComponents")
            .setMaster("spark://master:7077")
        val sc = new SparkContext(conf)

        val path = "file:///home/user/data/web-Google.txt"

        // Load the edges as a graph
        val graph: Graph[Int, Int] = GraphLoader.edgeListFile(
            sc,
            path,
            edgeStorageLevel = StorageLevel.MEMORY_AND_DISK,
            vertexStorageLevel = StorageLevel.MEMORY_AND_DISK
        )

        // Compute the weakly connected components
        val weaklyConnectedComponents = graph.connectedComponents().vertices.collect().map{
            case (id, comp) => s"$id $comp"
        }
        
        // Write the output to a file
        val outputs = new File("/home/user/workspace-graphx/outputs/weaklyConnectedComponents.txt")
        val bw = new BufferedWriter(new FileWriter(outputs))
        bw.write(weaklyConnectedComponents.mkString("\n"))
        bw.write("\n")
        bw.close()

        // Stop the Spark context
        sc.stop()
    }
}
